Doc/reference/expressions.rst

```diff
@@ -1687,6 +1687,11 @@ Or, when processing a file stream in chunks:
    while chunk := file.read(9000):
        process(chunk)
 
+Assignment expressions must be surrounded by parentheses when used
+as sub-expressions in subscript, slicing, conditional, lambda,
+keyword-argument, generator, and comprehension-if expressions.  They
+can only be used as is in if and while statements and in decorators.
```

Comment:
```suggestion
can only be used unparenthesized in if and while statements and in decorators.
```

Reply:
Agreed.  In addition, a list of unparenthesized subexpressions would be shorter: positional arguments in calls and single-value-expressions in set and list displays, including comprehensions. If there is something I miss, the list is still safe.

Reply:
Can you just apply this so we can land it? You can file a new issue to allow more scenarios.

Also, can you double check that the difference for () and [] is not a regression in the new parser?

Reply:
This is indeed a regression. I'll have a look at this as well.

Reply:
I see this in 3.8.6 as well as 3.10, so it is an oddity but not a regression.
```
>>> (a:=1, b:=2)
(1, 2)
>>> a:=1, b:=2
SyntaxError: invalid syntax

Reply:
That's a different thing that what I was asking about -- I was asking about
```
[a := i for i in ...]
(a := i for i in ...)
```
The former is valid in 3.8 through 3.10, the latter in 3.8 but not in later versions (until @lysnikolaou fixed it earlier today :-). So that's definitely a regression.

The thing you show is due to a generalization of the rule that `:=` is not allowed at the top level:
```
>>> (a := 1)
1
>>> a := 1
  File "<stdin>", line 1
    a := 1
          ^
SyntaxError: invalid syntax
```
I wouldn't want to change that.